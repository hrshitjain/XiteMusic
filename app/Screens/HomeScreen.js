import React from "react";
import {View, FlatList, Text} from 'react-native'
import RowItem from "../Components/RowItem";
import { getVideoLabels } from "../Services/APIService";
import styles from "../Styles/HomeStyles";
import SearchBar from "../Components/Searchbar";
import Icon from 'react-native-vector-icons/Ionicons';
import PopUp from "../Components/PopUp";

export default class HomeScreen extends React.Component{
    constructor(){
        super();
        this.state = {
            orginalData: [],
            labels:[],
            genres:[],
            isPopupVisible:false,
            selectedValues: [],
            searchText:''
        }
    }
    componentDidMount(){
        this.fetchLabels();
    }
    fetchLabels = async() =>{
        const response = await getVideoLabels();
        if(response != null){
            this.setState({
                orginalData:response.videos,
                labels:response.videos,
                genres: response.genres
            })
        }
    }
    renderRow = (item) =>{
        return <RowItem name={item.title}
                        artist={item.artist}
                        image={item.image_url}/>;
    }
    emptyComponent = () => {
        return(
            <View style={styles.emptyListContainer}>
                <Icon name="cloud-offline"
                      style={styles.emptyListIcon}
                      color="white"
                      size={180}/>
                <Text style={styles.emptyListText}>We could'nt get the videos for you this time, but we are working on it!!</Text>
            </View>
        )
    }
    setPopupVisibility = (visibility) =>{
        this.setState({
            isPopupVisible: visibility
        })
    }
    toggleGenre = (itemid) =>{
        const { selectedValues } = this.state;
        const index = selectedValues.indexOf(itemid);
        if(index > -1){
            selectedValues.splice(index, 1);
        }
        else{
            selectedValues.push(itemid);
        }
        this.setState({
          selectedValue: {...selectedValues},
        })
    }
    setSearchPhrase = (searchText) =>{
        const filteredData = this.searchFilter(searchText); 
        const filteredLabels = this.genreFilter(filteredData);
        this.setState({
            isPopupVisible:false,
            labels: filteredLabels,
            searchText: searchText
        })
    }
    applyFilter = () =>{
        const filteredData = this.searchFilter(this.state.searchText) 
        const filteredLabels = this.genreFilter(filteredData);
        this.setState({
            isPopupVisible:false,
            labels: filteredLabels
        })
    }
    onPopupClose = () =>{
        this.setState({
            isPopupVisible:false,
        })
    }
    onClearFilter = () => {
        const filteredLabels = this.searchFilter(this.state.searchText);
        this.setState({
            isPopupVisible:false,
            labels: filteredLabels,
            selectedValues:[]
        })
    }
    searchFilter = (searchText) => {
        const {orginalData} = this.state;
        const filterText = searchText.toUpperCase();
        let filteredData;
        if(searchText){
            filteredData = orginalData.filter(label => 
            {
                return label.title.toString().toUpperCase().indexOf(filterText) > -1 
                        || label.artist.toString().toUpperCase().indexOf(filterText) > -1 ;
    
            });   
        }
        else{
            filteredData = orginalData;
        }
        return filteredData;   
    }
    genreFilter = (filteredData) =>{
        const {selectedValues} = this.state;
        let filteredLabels;
        if(selectedValues.length > 0){
            filteredLabels = filteredData.filter((label) => selectedValues.indexOf(label.genre_id) > -1);
        }else{
            filteredLabels = filteredData;
        }
        return filteredLabels;
    }
    render(){
        const currentTime = new Date();
        let greeting;
        if(currentTime.getHours() < 12){
            greeting = 'Good Morning !!'    
        }
        else{
            greeting = 'Good Evening !!'
        }
        return(
            <View style={{backgroundColor:'black'}}>
                <View style={styles.logoContainer}>
                    <Icon name="musical-notes"
                                    style={styles.logo}
                                    color="white"
                                    onPress={()=>this.setPopupVisibility(true)}
                                    size={40}/>
                    <Text style={styles.greeting}>{greeting}</Text>
                </View>
                
                <View style={styles.toolbar}>
                    <SearchBar setSearchPhrase={this.setSearchPhrase}  />
                    <Icon name="filter"
                          style={styles.filter}
                          color="white"
                          onPress={()=>this.setPopupVisibility(true)}
                          size={40}/>
                </View>
                <PopUp isVisible={this.state.isPopupVisible}
                       genres={this.state.genres}
                       onClear={this.onClearFilter}
                       applyFilter={this.applyFilter}
                       onClose={this.onPopupClose}
                       selectedValues={this.state.selectedValues}
                       onValueChange={this.toggleGenre}
                       setVisibility={this.setPopupVisibility}/>
                <FlatList data={this.state.labels}
                          renderItem={({item}) => this.renderRow(item)}
                          keyExtractor={item => item.id}
                          numColumns={2}
                          ListEmptyComponent={this.emptyComponent}
                          extraData={this.state.orginalData} />

            </View>
        );
    }
}   