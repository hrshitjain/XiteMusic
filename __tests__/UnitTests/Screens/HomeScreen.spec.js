import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import HomeScreen from '../../../app/Screens/HomeScreen';

describe('FilterRow component render', () =>{
    it('Snapshot test', () =>{
        const {toJSON} = render(<HomeScreen />);
        expect(toJSON()).toMatchSnapshot();
    });
    it('Root View has 4 children', () => {
        const {toJSON} = render(<HomeScreen />);
        expect(toJSON().children.length).toBe(4);
    });
    it('Greeting Label is rendered',()=>{
        const currentTime = new Date();
        let greeting;
        if(currentTime.getHours() < 12){
            greeting = 'Good Morning !!'    
        }
        else{
            greeting = 'Good Evening !!'
        }
        const {getAllByText} = render(<HomeScreen/>);
        expect(getAllByText(greeting).length).toBe(1);
    });
    it('Search bar is rendered',()=>{
        const {getAllByPlaceholderText} = render(<HomeScreen/>);
        expect(getAllByPlaceholderText('Artist, music ...').length).toBe(1);
    });
});

describe("Search functionality", () => {
    it('Search music albums by title', () =>{
        const {getByPlaceholderText, getByText} = render(<HomeScreen/>);
        const searchBar = getByPlaceholderText('Artist, music ...');
        fireEvent.changeText(searchBar, 'All In');
        expect(getByText('All In').length).toBe(1);
    })
})