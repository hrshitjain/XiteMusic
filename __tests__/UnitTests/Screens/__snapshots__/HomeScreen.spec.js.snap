// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterRow component render Snapshot test 1`] = `
<View
  style={
    Object {
      "backgroundColor": "black",
    }
  }
>
  <View
    style={
      Object {
        "alignContent": "center",
        "flexDirection": "row",
        "justifyContent": "flex-start",
      }
    }
  >
    <Text
      allowFontScaling={false}
      onPress={[Function]}
      selectable={false}
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 40,
          },
          Object {
            "alignSelf": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
          },
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 30,
          "fontWeight": "bold",
          "margin": 20,
        }
      }
    >
      Good Evening !!
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 20,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "margin": 10,
          "width": "80%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#d9dbda",
            "borderRadius": 15,
            "flexDirection": "row",
            "padding": 10,
            "width": "95%",
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "black",
                "fontSize": 20,
              },
              Object {
                "marginLeft": 1,
              },
              Object {
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <TextInput
          clearButtonMode="while-editing"
          onChangeText={[Function]}
          placeholder="Artist, music ..."
          style={
            Object {
              "fontSize": 20,
              "marginLeft": 10,
              "width": "90%",
            }
          }
        />
      </View>
    </View>
    <Text
      allowFontScaling={false}
      onPress={[Function]}
      selectable={false}
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 40,
          },
          Object {
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-around",
            "margin": 10,
          },
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
  <RCTScrollView
    ListEmptyComponent={[Function]}
    data={Array []}
    extraData={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginTop": 100,
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 180,
              },
              Object {
                "alignSelf": "center",
              },
              Object {
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 16,
              "margin": 20,
              "textAlign": "center",
            }
          }
        >
          We could'nt get the videos for you this time, but we are working on it!!
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
